File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/annotations/parser/filter/primary/PreCheckoutQueryFilterParser.java
package ru.ioleksiv.telegram.bot.core.controller.annotations.parser.filter.primary;

import org.springframework.stereotype.Component;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primary.PreCheckoutQueryFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.secondary.OrderInfoFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.secondary.UserFilter;
import ru.ioleksiv.telegram.bot.api.model.objects.payments.PreCheckoutQuery;
import ru.ioleksiv.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import ru.ioleksiv.telegram.bot.core.controller.annotations.parser.finder.Finder;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.Checker;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.impl.UnionExtractChecker;

import java.util.Optional;

@Component
public class PreCheckoutQueryFilterParser implements FilterParser<PreCheckoutQueryFilter, PreCheckoutQuery> {

    @Override
    public Checker<PreCheckoutQuery> createChecker(PreCheckoutQueryFilter annotation, Finder finder) {
        UnionExtractChecker<PreCheckoutQuery> unionExtractChecker = new UnionExtractChecker<>();

        UserFilter from = annotation.from();
        if (from.value().isActive()) {
            unionExtractChecker.add(in -> Optional.ofNullable(in.getFrom()), finder.find(from));
        }
        StringFilter currency = annotation.currency();
        if (currency.value().isActive()) {
            unionExtractChecker.add(in -> Optional.ofNullable(in.getCurrency()), finder.find(currency));
        }
        StringFilter totalAmount = annotation.totalAmount();
        if (totalAmount.value().isActive()) {
            unionExtractChecker.add(in -> Optional.ofNullable(in.getTotalAmount()), finder.find(totalAmount));
        }
        StringFilter invoicePayload = annotation.invoicePayload();
        if (invoicePayload.value().isActive()) {
            unionExtractChecker.add(in -> Optional.ofNullable(in.getInvoicePayload()), finder.find(invoicePayload));
        }
        StringFilter shippingOptionId = annotation.shippingOptionId();
        if (shippingOptionId.value().isActive()) {
            unionExtractChecker.add(in -> Optional.ofNullable(in.getShippingOptionId()), finder.find(shippingOptionId));
        }
        OrderInfoFilter orderInfo = annotation.orderInfo();
        if (orderInfo.value().isActive()) {
            unionExtractChecker.add(in -> Optional.ofNullable(in.getOrderInfo()), finder.find(orderInfo));
        }

        return unionExtractChecker;
    }

    @Override
    public Class<PreCheckoutQueryFilter> getAnnotationClass() {
        return PreCheckoutQueryFilter.class;
    }

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/controller/annotations/parser/filter/telegram/PreCheckoutQueryFilterParser.kt
package io.github.oleksivio.tl.kbot.core.controller.annotations.parser.filter.telegram

import io.github.oleksivio.tl.kbot.core.annotations.filter.composite.PreCheckoutQueryFilter
import io.github.oleksivio.tl.kbot.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.tl.kbot.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.tl.kbot.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.controller.handler.check.impl.UnionExtractValidator
import io.github.oleksivio.tl.kbot.server.api.objects.payments.PreCheckoutQuery

class PreCheckoutQueryFilterParser :
    FilterParser<PreCheckoutQueryFilter, PreCheckoutQuery> {

    override val parserAnnotationClass = PreCheckoutQueryFilter::class

    override fun createChecker(annotation: PreCheckoutQueryFilter, finder: Finder): Validator<PreCheckoutQuery> {
        val unionExtractValidator =
            UnionExtractValidator<PreCheckoutQuery>()

        annotation.validator
            .map { validatorName -> finder.find(validatorName, PreCheckoutQuery::class) }
            .forEach { validator -> unionExtractValidator.add({ it }, validator) }

        val from = annotation.from
        if (from.status.isActive) {
            unionExtractValidator.add({ it.from }, finder.find(from))
        }
        val currency = annotation.currency
        if (currency.status.isActive) {
            unionExtractValidator.add({ it.currency }, finder.find(currency))
        }
        val totalAmount = annotation.totalAmount
        if (totalAmount.status.isActive) {
            unionExtractValidator.add({ it.totalAmount }, finder.find(totalAmount))
        }
        val invoicePayload = annotation.invoicePayload
        if (invoicePayload.status.isActive) {
            unionExtractValidator.add({ it.invoicePayload }, finder.find(invoicePayload))
        }
        val shippingOptionId = annotation.shippingOptionId
        if (shippingOptionId.status.isActive) {
            unionExtractValidator.add({ it.shippingOptionId }, finder.find(shippingOptionId))
        }
        val orderInfo = annotation.orderInfo
        if (orderInfo.status.isActive) {
            unionExtractValidator.add({ it.orderInfo }, finder.find(orderInfo))
        }

        return unionExtractValidator
    }
}

